transformations/tests/test_block_index_inject.py

```diff
@@ -0,0 +1,327 @@
+
+    yield workdir, request.param
+
+    rmtree(workdir)
+
+
+@pytest.mark.parametrize('frontend', available_frontends(xfail=[(OMNI,
+                         'OMNI fails to import undefined module.')]))
+def test_blockview_to_fieldview_pipeline(horizontal, blocking, config, frontend, blockview_to_fieldview_code):
+
+    config['routines'] = {
+        'driver': {'role': 'driver'}
+    }
+
+    scheduler = Scheduler(
+        paths=(blockview_to_fieldview_code[0],), config=config, seed_routines='driver', frontend=frontend
+    )
+    scheduler.process(BlockViewToFieldViewTransformation(horizontal, global_gfl_ptr=True))
+    scheduler.process(BlockIndexInjectTransformation(blocking))
+
+    kernel = scheduler['#kernel'].ir
+    aliased_bounds = not blockview_to_fieldview_code[1]
+    ibl_expr = blocking.index
+    if aliased_bounds:
+        ibl_expr = blocking.index_expressions[1]
+
+    assigns = FindNodes(Assignment).visit(kernel.body)
+
+    # check that access pointers for arrays without horizontal index in dimensions were not updated
+    assert assigns[0].lhs == f'ydvars%var%p_field(:,:,{ibl_expr})'
+    assert assigns[1].lhs == f'ydvars%var%p_field(:,:,{ibl_expr})'
+
+    # check that vector notation was resolved correctly
+    assert assigns[2].lhs == f'yda_data%p_field(jl, :, {ibl_expr})'
+    assert assigns[3].lhs == f'ydvars%var%p_field(jl, :, {ibl_expr})'
+
+    # check thread-local ydvars%gfl_ptr was replaced with its global equivalent
+    gfl_ptr_vars = {v for v in FindVariables().visit(kernel.body) if 'ydvars%gfl_ptr' in v.name.lower()}
+    gfl_ptr_g_vars = {v for v in FindVariables().visit(kernel.body) if 'ydvars%gfl_ptr_g' in v.name.lower()}
+    assert gfl_ptr_g_vars
+    assert not gfl_ptr_g_vars - gfl_ptr_vars
+
+    assert assigns[4].lhs == f'ydvars%gfl_ptr_g(jfld)%ptr%p_field(jl,:,{ibl_expr})'
+    assert assigns[4].rhs == f'yda_data%p_field(jl,:,{ibl_expr})'
+    assert assigns[5].lhs == f'container%vars(ydvars%gfl_ptr_g(jfld)%comp)%p_field(jl,:,{ibl_expr})'
+
+    # check callstatement was updated correctly
+    call = FindNodes(CallStatement).visit(kernel.body)[0]
+    assert f'yda_data%p_field(:,:,{ibl_expr})' in call.arguments
+
+
+@pytest.mark.parametrize('frontend', available_frontends(xfail=[(OMNI,
+                         'OMNI correctly complains about rank mismatch in assignment.')]))
+def test_simple_blockindex_inject(blocking, frontend):
+    fcode = """
+subroutine kernel(nlon,nlev,nb,var)
+  implicit none
+
+  interface
+    subroutine compute(nlon,nlev,var)
+      implicit none
+      integer, intent(in) :: nlon,nlev
+      real, intent(inout) :: var(nlon,nlev)
+    end subroutine compute
+  end interface
+
+  integer, intent(in) :: nlon,nlev,nb
+  real, intent(inout) :: var(nlon,nlev,nb) !... this dummy arg was potentially promoted by a previous transformation
+
+  integer :: ibl
+
+  do ibl=1,nb !... this loop was potentially lowered by a previous transformation
+     var(:,:) = 0.
```

Comment:
Just quickly skimmed through the PR so I haven't yet understood how everything works. However, this is not valid Fortran, is it? `var` is defined as 3d array and you access the array as it would be 2d?!

Reply:
Yes I split it in two transformations so that one can have the CPG/SL specific logic and the second can be more general which can potentially be used in other transformation pipelines.